File: submissions/online-judges/cf/cf-gym/asia-taiwan-online-programming-contest-2023/E.py
a, b, m = map(int, input().split())

A = [[a, m - 1], [1, 0]]
identity = [[1, 0], [0, 1]]


def mul(a, b):
    sz = len(a)
    d = [[0 for _ in range(sz)] for _ in range(sz)]
    for i in range(sz):
        for j in range(sz):
            for k in range(sz):
                d[i][j] = (d[i][j] + a[i][k] * b[k][j]) % m
    return d


def fpow(a, b):
    ret = identity
    while (b > 0):
        if (b % 2 == 1):
            ret = mul(ret, a)
        a = mul(a, a)
        b = b // 2
    return ret


Ab = fpow(A, b - 1)
ret = (Ab[0][0] * a + Ab[0][1] * 2) % m
if ret < 0:
    ret += m

print(ret)

File: submissions/online-judges/cf/cf-gym/asia-taiwan-online-programming-contest-2023/E.cpp

#include <bits/stdc++.h>
using namespace std;

using ll = long long;
using ld = long double;

const int INF = 2e9;

#include <bits/stdc++.h>
using namespace std;

using ll = long long;
using matrix = vector<vector<ll>>;

matrix identity;
ll mod;

matrix mul(matrix a, matrix b) {
    int sz = (int)a.size();
    matrix d(sz, vector<ll>(sz));
    for (int i = 0; i < sz; i++) {
        for (int j = 0; j < sz; j++) {
            for (int k = 0; k < sz; k++) {
                d[i][j] = (d[i][j] + a[i][k] * b[k][j]) % mod;
            }
        }
    }
    return d;
}

matrix fpow(matrix a, ll b) {
    matrix ret = identity;
    while (b) {
        if (b & 1) ret = mul(ret, a);
        a = mul(a, a);
        b >>= 1;
    }
    return ret;
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0), cout.tie(0);

    ll a, b;
    cin >> a >> b >> mod;
    a %= mod;
    identity = {{1, 0}, {0, 1}};
    matrix A = {{a, -1}, {1, 0}};

    int sz = 3;
    for (int i = 0; i < sz; i++) {
        vector<ll> temp;
        for (int j = 0; j < sz; j++) {
            temp.push_back((i == j) ? 1 : 0);
        }
        identity.push_back(temp);
    }

    printMatrix(identity);
}

void el_psy_congroo() {
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    int TC = 1;
    cin >> TC;
    for (int i = 1; i <= TC; i++) {
        el_psy_congroo();
    }

    return 0;
}